client/my-sites/plugins/marketplace/marketplace-domain-upsell/index.tsx

```diff
@@ -0,0 +1,204 @@
+import type { DomainSuggestions } from '@automattic/data-stores';
+
+/**
+ * Internal dependencies
+ */
+import CalypsoShoppingCartProvider from 'calypso/my-sites/checkout/calypso-shopping-cart-provider';
+import { domainRegistration } from 'calypso/lib/cart-values/cart-items';
+import {
+	MARKETPLACE_FLOW_ID,
+	ANALYTICS_UI_LOCATION_MARKETPLACE_DOMAIN_SELECTION,
+} from 'calypso/my-sites/plugins/marketplace/constants';
+import { getSelectedSite } from 'calypso/state/ui/selectors';
+import { getProductsList, isProductsListFetching } from 'calypso/state/products-list/selectors';
+import { fillInSingleCartItemAttributes } from 'calypso/lib/cart-values';
+import Masterbar from 'calypso/layout/masterbar/masterbar';
+import Item from 'calypso/layout/masterbar/item';
+import getPreviousPath from 'calypso/state/selectors/get-previous-path';
+import { HorizontalRule } from '@wordpress/components';
+import ExternalLink from 'calypso/components/external-link';
+import MarketplaceShoppingCart from 'calypso/my-sites/plugins/marketplace/components/marketplace-shopping-cart';
+import theme from 'calypso/my-sites/plugins/marketplace';
+
+/**
+ * Style dependencies
+ */
+import 'calypso/my-sites/plugins/marketplace/marketplace-domain-upsell/style.scss';
+
+const MarketplaceHeaderTitle = styled.h1`
+	font-size: 2rem;
+	margin-bottom: 8px;
+`;
+const MarketplaceHeaderSubTitle = styled.h2`
+	font-size: 0.875rem;
+	margin-bottom: 15px;
+`;
+
+function MarketplaceDomainUpsellHeader() {
+	return (
+		<div>
+			<MarketplaceHeaderTitle className="marketplace-domain-upsell__title wp-brand-font">
+				{ translate( 'Choose a domain' ) }
+			</MarketplaceHeaderTitle>
+			<MarketplaceHeaderSubTitle>
+				{ translate(
+					'Yoast SEO requires a top level domain to index your site and help you rank higher.'
+				) }
+			</MarketplaceHeaderSubTitle>
+			<ExternalLink
+				href="https://yoast.com/domain-names-seo"
+				target="_blank"
+				rel="noopener noreferrer"
+				icon
+			>
+				{ translate( 'Domain names and their influence on SEO' ) }
+			</ExternalLink>
+			<HorizontalRule />
+		</div>
+	);
+}
+
+const getSiteNameFromURL = ( url: string ) => url?.split( '.' )?.[ 0 ] ?? url;
+
+function CalypsoWrappedMarketplaceDomainUpsell(): JSX.Element {
+	const [ selectedDomainProductUUID, setDomainProductUUID ] = useState< string >( '' );
+	const [ selectedDomain, setDomain ] = useState< DomainSuggestions.DomainSuggestion | null >(
+		null
+	);
+
+	const [ isExpandedBasketView, setIsExpandedBasketView ] = useState( false );
+	const { addProductsToCart, replaceProductsInCart, removeProductFromCart } = useShoppingCart();
+	const products = useSelector( getProductsList );
+	const previousPath = useSelector( getPreviousPath );
+	const isFetchingProducts = useSelector( isProductsListFetching );
+	const selectedSite = useSelector( getSelectedSite );
+	const { wpcom_url } = selectedSite;
+	const siteName = getSiteNameFromURL( wpcom_url );
```

Comment:
I dug a bit and found that `wpcom_url` isn't always available. From my observation, it will be there if a site has a custom domain set as the primary domain. On the other hand, this field won't be there and we will need to implement a fallback strategy, probably based on `site_slug`: 
<img width="517" alt="圖片" src="https://user-images.githubusercontent.com/1842898/117238070-d0e8ab00-ae5e-11eb-9fa1-4435707404f6.png">

I know I said that let's implement this as a follow-up. However, that was because I thought the related code wasn't there. Since it's there, let's either fix it or remove it completely. The bottom line is that we shouldn't ever commit something broken or redundant.

Reply:
Oh so that was the reason. I will fix this right away! :)

Reply:
I did a small experimentation and seems like this data structure does not reliably contain the wpcom domain in it. So I went ahead and explored how this data is extracted in the domain management page. I have created some new selectors for future use of others who want to use this feature as well.

Changes are done in this commit 3bae69a6b14385ff3f1c69475b6e4a7e5914fd1b